thrift/src/main/java/com/intellij/plugins/thrift/lang/ThriftFoldingBuilder.java

```diff
@@ -0,0 +1,178 @@
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.plugins.thrift.lang.lexer.ThriftElementType;
+import com.intellij.psi.JavaRecursiveElementWalkingVisitor;
+import com.intellij.psi.PsiElement;
+import com.intellij.plugins.thrift.lang.psi.*;
+import com.intellij.psi.tree.IElementType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.intellij.plugins.thrift.lang.lexer.ThriftTokenTypes.*;
+
+
+public class ThriftFoldingBuilder extends FoldingBuilderEx implements DumbAware {
+  @Override
+  public FoldingDescriptor @NotNull [] buildFoldRegions(@NotNull PsiElement root, @NotNull Document document, boolean quick) {
+    List<FoldingDescriptor> descriptors = new ArrayList<>();
+    root.accept(new ThriftFoldingRecursiveElementWalkingVisitor(descriptors));
+    return descriptors.toArray(FoldingDescriptor.EMPTY_ARRAY);
+  }
+
+  @Override
+  public @Nullable String getPlaceholderText(@NotNull ASTNode node) {
+    PsiElement psiElement = node.getPsi();
+    String placeholderText = "...";
+    if (psiElement instanceof ThriftEnum) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftSenum) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftStruct) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftUnion) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftException) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftService) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftXsdAttrs) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftConstMap) {
+      placeholderText = "{...}";
+    } else if (psiElement instanceof ThriftConstList) {
+      placeholderText = "[...]";
+    } else if (isOfThriftElementType(node.getElementType(), BLOCKCOMMENT)) {
+      placeholderText = "/*...*/";
+    }
+    return placeholderText;
+  }
+
+  @Override
+  public boolean isCollapsedByDefault(@NotNull ASTNode node) {
+    return false;
+  }
+
+  static boolean isOfThriftElementType(IElementType elementType, IElementType thriftElementType) {
+    if (elementType instanceof ThriftElementType) {
+      return elementType.equals(thriftElementType);
+    }
+    return false;
+  }
+
+  static class ThriftFoldingRecursiveElementWalkingVisitor extends JavaRecursiveElementWalkingVisitor {
+    private final List<FoldingDescriptor> descriptors;
+
+    public ThriftFoldingRecursiveElementWalkingVisitor(List<FoldingDescriptor> descriptors) {
+      this.descriptors = descriptors;
+    }
+
+    @Override
+    public void visitElement(@NotNull PsiElement element) {
+      super.visitElement(element);
+
+      if (element instanceof ThriftEnum) {
+        this.visitThriftEnum((ThriftEnum) element);
```

Comment:
Why not to call `visitElementWithCurlyBraceChildren` directly for all elements that have curly braces?